### ISSUE TITLE
CE2: IO's parallel applicative isn't commutative

### ISSUE THREAD

**Issue by kubukoz** [typelevel/cats-effect#1293]
An example speaks a thousand words:

```scala
import cats.Applicative
import cats.effect.ExitCode
import cats.effect.IO
import cats.effect.IOApp
import cats.effect.concurrent.Deferred
import cats.implicits._
import cats.effect.implicits._
import cats.effect.Bracket

object Demo extends IOApp {
  // Like .uncancelable, but stronger
  def uncancelableWait[F[_]: Bracket[*[_], E], A, E](fa: F[A]): F[A] = fa.bracket(_.pure[F])(_ => Applicative[F].unit)

  import scala.concurrent.duration._

  def run(args: List[String]): IO[ExitCode] =
    Deferred[IO, Unit].flatMap { latch =>
      val left: IO[Unit] = uncancelableWait(latch.get)
      val right: IO[Unit] = IO.never.onCancel(latch.complete(()))

      //this hangs
      left &> right
      //this completes after sleep wins
      // left &> right
    }.race(IO.sleep(500.millis)).as(ExitCode.Success)

}
```

What's happening here:

- `left` is uncancelable (`cancel` will wait for its full completion, which is when the latch is completed)
- `right` never terminates, but completes the latch when canceled
- a result of zipping them in parallel is raced with (anything really)
- when the other side wins (it must - `left` can't complete unless the whole thing is canceled), `right`'s cancelation handling should complete the latch and, as a result, `left`. This doesn't happen if `left` is on the LHS of `&>`.

I didn't check yet whether `right`'s `onCancel` actually starts - I suppose not, since if it did, `left` would also finish...

**Comment by kubukoz** [comment]
As far as I can see without doing some deeper debugging, `IOParMap`'s implementation calls `pushPair` on the `IOConnection`, which pushes `CancelUtils.cancelAll(lh.cancel, rh.cancel)` to the stack of finalizers. `cancelAll` runs the finalizers in order, so that would explain the behavior.

If we can bend the rules and say order of finalizers doesn't need to obey commutativity, this would be fine, but I'm not sure about that.

Is there any reason why we couldn't run these finalizers in parallel?

**Comment by djspiewak** [comment]
> Is there any reason why we couldn't run these finalizers in parallel?

None. They should be
